pkg/cloudprovider/provider/vsphere/helper.go

```diff
@@ -218,13 +211,80 @@ func createClonedVM(ctx context.Context, vmName string, config *Config, session
+	// Based on https://github.com/vmware/govmomi/blob/v0.22.1/govc/vm/clone.go#L358
+	if config.DatastoreCluster != "" && config.Datastore == "" {
+		klog.Infof("Choosing initial datastore placement for vm %s from datastore cluster %s", vm.Name(), config.DatastoreCluster)
+		storagePod, err := session.Finder.DatastoreCluster(ctx, config.DatastoreCluster)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get datastore cluster: %v", err)
+		}
+
+		// Build pod selection spec from config spec
+		podSelectionSpec := types.StorageDrsPodSelectionSpec{
+			StoragePod: types.NewReference(storagePod.Reference()),
+		}
+
+		// Get the virtual machine reference
+		vmref := vm.Reference()
+
+		// TODO(irozzo) moveAllDiskBackingsAndConsolidate does not seem to work with RecommendDatastore,
+		// try to better understand the reason and the implications.
+		// https://code.vmware.com/docs/4206/vsphere-web-services-api-reference/doc/vim.vm.RelocateSpec.DiskMoveOptions.html
+		cloneSpec.Location.DiskMoveType = string(types.VirtualMachineRelocateDiskMoveOptionsMoveAllDiskBackingsAndDisallowSharing)
+		// Build the placement spec
+		storagePlacementSpec := types.StoragePlacementSpec{
+			Folder:           types.NewReference(folder.Reference()),
+			Vm:               &vmref,
+			CloneName:        vm.Name(),
+			CloneSpec:        cloneSpec,
+			PodSelectionSpec: podSelectionSpec,
+			Type:             string(types.StoragePlacementSpecPlacementTypeClone),
+		}
+
+		// Get the storage placement result
+		storageResourceManager := object.NewStorageResourceManager(session.Client.Client)
+		result, err := storageResourceManager.RecommendDatastores(ctx, storagePlacementSpec)
+		if err != nil {
+			return nil, fmt.Errorf("error occurred while getting storage placement recommendation: %v", err)
+		}
+
+		// Get the recommendations
+		recommendations := result.Recommendations
+		if len(recommendations) == 0 {
+			return nil, fmt.Errorf("no recommendations")
+		}
+
+		// Get the first recommendation
+		ds := recommendations[0].Action[0].(*types.StoragePlacementAction).Destination.Reference()
+		return &ds, nil
+	} else if config.DatastoreCluster == "" && config.Datastore != "" {
+		datastore, err := session.Finder.Datastore(ctx, config.Datastore)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get datastore: %v", err)
+		}
+		return types.NewReference(datastore.Reference()), nil
+	} else {
+		return nil, fmt.Errorf("please provide either a datastore or a datastore cluster")
 	}
+}
+
+func uploadAndAttachISO(ctx context.Context, session *Session, vmRef *object.VirtualMachine, localIsoFilePath string) error {
 	p := soap.DefaultUpload
 	remoteIsoFilePath := fmt.Sprintf("%s/%s", vmRef.Name(), "cloud-init.iso")
+	var props mo.VirtualMachine
+	// Obtain VM properties
+	if err := vmRef.Properties(ctx, vmRef.Reference(), nil, &props); err != nil {
```

Comment:
There was a parameter for this function where the datastore name is sent via the machine config, thus this is not needed. Is there a reason why removed that and fetched the properties of the vm to get the ds path, while you had before hand?

Reply:
I could pass the datastore explicitly to save an API call as a small optimisation, but as it is a property of the VM there is no real need of passing it explicitly.

Reply:
You are adding a boilerplate to achieve something that you already have. Please retain the code as it was because there is no real value in executing a request to vsphere to request for something you have beforehand!

Reply:
it's more self contained now, the information about the datastore is already associated to the VM. Passing it explicitly is more effective because we save an API call, but is redundant. On the other hand if we want to optimise we can directly pass the reference to the datastore object, it's even better than what we had before. I don't think that this kind of optimisation should be the main driver at this stage.

Reply:
The information of the vm should come from the machine configuration and the machine configuration is coming from the MachineDeployment configurations and those configurations are the source of truth which the vm should be bounded with!!

Reply:
the `MachineDeployment` after my development may not contain the datastore